### ISSUE TITLE
Multiprocess/Multithreading Debug

### ISSUE THREAD

**Issue by jiaqifan123** [cyyever/distributed_learning_simulator#5]
![%W)P)2 24JWU$26NP@2A43](https://github.com/cyyever/distributed_learning_simulator/assets/150891844/b4444b08-e6d1-4892-9062-055d2454b310)
Hello author，
When I debug this part of the code, the process pool submits the task function and starts training directly, I have no way to see the algorithm execution process and the model training process.how I should debug it? Thank you very much for your answer.

**Comment by cyyever** [comment]
What do you mean by "the algorithm execution process and the model training process", the training will output local validation loss and training loss. And which debug info do you want?

**Comment by jiaqifan123** [comment]
I want to see how the algorithm is executed step by step, rather than executing this part of the code and directly exporting the loss. Such as training on the client side, through the neural network, model aggregation. I debug this part of the code, there is no way to enter the algorithm inside, can not intuitively see the details inside. Thank you for your answer

> What do you mean by "the algorithm execution process and the model training process", the training will output local validation loss and training loss. And which debug info do you want?

**Comment by cyyever** [comment]
The source code for training is in https://github.com/cyyever/torch_toolbox/blob/main/cyy_torch_toolbox/trainer.py.  I don't want to output detailed log for local clients because in large-scale FL consisting of over 100 clients it will generate too many lines of output.

**Comment by cyyever** [comment]
For FEDAVG implementation you can see the code in  distributed_learning_simulator/algorithm/fed_avg_algorithm.py

**Comment by jiaqifan123** [comment]
Hi author，
In the training.py file, when I debug to the "submit function" in line 105, it directly executes the "NNADQ algorithm," and when I debug to the "submit function" in line 113, it directly trains the local model. How should I track the "NNADQ algorithm" execution process in line 105 and the "local model training, OBD algorithm and upload to server side for model aggregation" process in line 113? Thank you very much for your answer.
![(I63PQXEH7H{1AH2DI~O)NG](https://github.com/cyyever/distributed_learning_simulator/assets/150891844/c69b86ad-1da0-431d-b33c-ee409c7f39ab)

**Comment by jiaqifan123** [comment]
Hi author，
In the training.py file, when I debug to the "submit function" in line 105, it executes the "NNADQ algorithm" directly, and when I debug to the "submit function" in line 113, it outputs the training loss directly. How should I track the execution process of the "NNADQ algorithm" in line 105 and the process of the algorithm in line 113, such as the training of the local model, the OBD algorithm, and the process of uploading to the server for model aggregation? Thank you very much for your answer.
![Uploading (I63PQXEH7H{1AH2DI~O)NG.png…]()

**Comment by jiaqifan123** [comment]
Hi author，
In the training.py file, when I debug to the "submit function" in line 105, it executes the "NNADQ algorithm" directly, and when I debug to the "submit function" in line 113, it outputs the training loss directly. How should I track the execution process of the "NNADQ algorithm" in line 105 and the process of the algorithm in line 113, such as the training of the local model, the OBD algorithm, and the process of uploading to the server for model aggregation? Thank you very much for your answer.
![(I63PQXEH7H{1AH2DI~O)NG](https://github.com/cyyever/distributed_learning_simulator/assets/150891844/8e86e070-24ab-49e7-ac79-39ecfec03b49)

**Comment by cyyever** [comment]
Due to configuration efficiency, I use factor pattern to encapsulate FL algorithms so it is harder to trace using a debugger. It is more easier to point out the source files. For NNADQ, the quantization algorithm is called at topology/quantized_endpoint.py and the implementation is located at https://github.com/cyyever/torch_algorithm/blob/main/cyy_torch_algorithm/quantization/deterministic.py

**Comment by cyyever** [comment]
It is easier to grep keywords to quickly find out related source code.
